client/src/shared/state/Inventory.js

```diff
@@ -104,16 +107,155 @@ class Inventory {
         if (!ItemTypes[itemConfig.typeCode].hasUseEffect) return;
 
         this.hotbar.push(itemConfig);
-
+        this.keyToSlotIndex[this.hotbar.length - 1] = itemConfig.slotIndex;
         PubSub.publish(HOTBAR_ITEM);
         PubSub.publish(MODIFY_INVENTORY_ITEM);
+        if (this.saveSession === true) {
+            this.saveHotbar();
+        }
+    }
+
+    resetMapping() {
+        Utils.warning("This is a bug found or a localStorage mismatch");
+        this.keyToSlotIndex = {};
+        this.saveHotbar();
+        this.hotbar = [];
+        this.defaultHotBar();
     }
 
-    removeFromHotbar(itemConfig) {
+    updateSlotIndex(slotIndex) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (this.keyToSlotIndex[key] >= slotIndex) {
+                const updateValue = this.keyToSlotIndex[key] - 1;
+                if (updateValue >= 0) {
+                    this.keyToSlotIndex[key] = updateValue;
+                    if (this.saveSession === true) {
+                        this.saveHotbar();
+                    }
+                }
+            }
+        });
+    }
+
+    updateMapHotBarKeys(evictKey) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (key >= evictKey) {
+                const updateKey = key - 1;
+                if (updateKey >= 0) {
+                    this.keyToSlotIndex[updateKey] = this.keyToSlotIndex[key];
+                    delete this.keyToSlotIndex[key];
+                }
+            }
+        });
+    }
+
+    updateMapSlotIndex(evictValue) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (this.keyToSlotIndex[key] >= evictValue) {
+                const newValue = this.keyToSlotIndex[key] - 1;
+                if (newValue >= 0) {
+                    this.keyToSlotIndex[key] = newValue;
+                }
+            }
+        });
+    }
+
+    removeFromHotbar(itemConfig, removedInventory = false) {
+        const getKeys = Object.keys(this.keyToSlotIndex);
+        const evictKey = getKeys.find((key) => this.keyToSlotIndex[key] === itemConfig.slotIndex);
+        const evictValue = this.keyToSlotIndex[evictKey];
+        if (typeof evictKey === "undefined") {
+            this.updateSlotIndex(itemConfig.slotIndex);
+            return;
+        }
+        delete this.keyToSlotIndex[evictKey];
+        this.updateMapHotBarKeys(evictKey);
+        if (removedInventory) {
+            this.updateMapSlotIndex(evictValue);
+        }
         this.hotbar = this.hotbar.filter((eachItem) => eachItem !== itemConfig);
 
         PubSub.publish(HOTBAR_ITEM);
         PubSub.publish(MODIFY_INVENTORY_ITEM);
+        if (this.saveSession === true) {
+            this.saveHotbar();
+        }
+    }
+
+    saveHotbar() {
+        const sessionIDMap = "saveHotBar";
+        const localStorageID = this.userName + sessionIDMap;
```

Comment:
Even if they did use the userName for another websites, the chance that it's gonna be "bbb" + "saveHotBar" --> "bbbsaveHotBar" is soo small, but if you want I can hash it, and store the hash as the unique idetifer.

Reply:
identical keys clashing isn't really the concern, it's more for if someone were to inspect their localstorage and they see "bbbsaveHotBar" then there's a chance they wouldn't know which website it came from. It would be better to have some way to identify the origin, like `dungeonz_bbb_saveHotBar` (then for other localstorage items dungeonz would follow the convention `dungeonz_bbb_blahblah`). But, the owners of dungeonz would have a better idea than me on what this convention would be (or is, if they already have established it)

Reply:
Or alternatively I can concatenate userName + (16 bit random hash) --> Gaurantee to be unique regardless of what website

Reply:
![image](https://user-images.githubusercontent.com/43216310/145159369-9a1b4f4e-022b-49aa-bb7e-42fe165db164.png)
This is an example of how they did it.

Reply:
I see, that's bad key use on them then I guess 🤷🏼‍♀️